File: app/src/main/java/com/concrete/android/challenge/di/module/ViewModelModule.java
package com.concrete.android.challenge.di.module;

import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import com.concrete.android.challenge.ViewModelFactory;
import com.concrete.android.challenge.ui.main.MainViewModel;
import com.concrete.android.challenge.ui.pull_request.PullRequestViewModel;
import com.concrete.android.challenge.ui.repository.RepositoryViewModel;
import com.concrete.android.challenge.ui.splash.SplashViewModel;
import dagger.Binds;
import dagger.Module;
import dagger.multibindings.IntoMap;

/**
 * @author Thiago Corredo
 * @since 2019-05-28
 */
@Module
public abstract class ViewModelModule {

  @Binds
  @IntoMap
  @ViewModelKey(PullRequestViewModel.class)
  abstract ViewModel bindPullRequestViewModel(PullRequestViewModel pullRequestViewModel);

  @Binds
  @IntoMap
  @ViewModelKey(RepositoryViewModel.class)
  abstract ViewModel bindRepositoryViewModel(RepositoryViewModel repositoryViewModel);

  @Binds
  @IntoMap
  @ViewModelKey(MainViewModel.class)
  abstract ViewModel bindMainViewModel(MainViewModel mainViewModel);

  @Binds
  @IntoMap
  @ViewModelKey(SplashViewModel.class)
  abstract ViewModel bindSplashViewModel(SplashViewModel splashViewModel);

  @Binds
  abstract ViewModelProvider.Factory bindsViewModelFactory(ViewModelFactory viewModelFactory);
}

File: app/src/main/java/com/ccortez/desafioconcrete/di/module/ViewModelModule.kt
package com.ccortez.desafioconcrete.di.module

import androidx.lifecycle.ViewModel
import com.ccortez.desafioconcrete.ui.main.MainViewModel
import com.ccortez.desafioconcrete.ui.main.RepositoryViewModel
import com.ccortez.desafioconcrete.utils.ViewModelKey
import dagger.Binds
import dagger.Module
import dagger.multibindings.IntoMap

@Module
abstract class ViewModelModule {

    @Binds
    @IntoMap
    @ViewModelKey(MainViewModel::class)
    abstract fun bindMainViewModel(viewModel: MainViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(RepositoryViewModel::class)
    abstract fun bindRepositoryViewModel(repositoryViewModel: RepositoryViewModel): ViewModel

}